File: kotlin/Lib/313.SuperUglyNumber.kt
// 313. Super Ugly Number

class LeetCode313SuperUglyNumber {
    private data class PrimePointer(val value: Int, var pos: Int)

    fun nthSuperUglyNumber(n: Int, primes: IntArray): Int {
        if (n == 1) return 1

        val pps = primes.map { PrimePointer(it, 0) }.toTypedArray()
        var dp = `_GenInitTable`(n)

        for (i in 0 until n) {
            dp[i] = _GetNextUglyNumber(dp, pps)
        }

        return dp[n-1]
    }

    private fun _GenInitTable(num: Int): IntArray {
        var dp = IntArray(num) {0}
        dp[0] = 1
        return dp
    }

    private fun _GetNextUglyNumber(dp: IntArray, pps: Array<PrimePointer>): Int {
        var uglyNumber = pps.map { it.value * dp[it.pos] }.min()!!

        for (p in pps) {
            if (uglyNumber == p.value * dp[p.pos])
                ++p.pos
        }

        return uglyNumber
    }
}

File: csharp/csharp/Lib/313.SuperUglyNumber.cs
/*
313. Super Ugly Number
*/

using System.Linq;

public class SuperUglyNumber {
	private class PrimePointer {
		public int Value;
		public int Pointer;
	}

    public int NthSuperUglyNumber(int n, int[] primes) {
		if (n == 1)
			return 1;
		
		var pps = _GenPrimePointers(primes);
		var dp = new int[n];
		dp[0] = 1;

		for (int i = 1; i < dp.Length; ++i) {
			dp[i] = pps.Min(p => p.Value * dp[p.Pointer]);

			foreach (var p in pps)
				if (dp[i] == p.Value * dp[p.Pointer])
					++p.Pointer;
		}

		return dp[n-1];
    }

	private PrimePointer[] _GenPrimePointers(int[] primes) {
		return primes.Select(p =>
			new PrimePointer {
				Value = p,
				Pointer = 0,
			}).ToArray();
	}
}